## Table of Contents

- Introduction
- Business Problem
- Objectives
- Data Understanding
- Data Preparation

## 1. INTRODUCTION

### In the telecommunications industry , alot of factors contributre to customer churn. SAome of the factors include, quality of service provison,  customer  service effectivceness, and competion from service providers. Employing predictive models such as logistice regression  and decission trees can enhance the sevice providers to easily anticipate churt and  put in place measures to conteract loss of clients.

## 2. BUSINESS PROBLEM

#### Customer retention is one of the key factor in  growth sustenance for any telecomunication  firm. Therefore customer churn; Aphenomena whereby a customer stops to utilize  company services,stands to be one of   greatest threat for any telecom firm. This study  explores predictive analysis to explore various factors contributing to customer churn. As well as employing data-driven approach to predict likelihood of customers churn based on various characteristics.

## 3. OBJECTIVES

#### 1. Understand the customer churn rate

#### 2. Identify the most influential factors contributing to customer churn

#### 3. Develop a predictive model that can accurately predict the likelihood of customers churn based on their past behavior and characteristics.



## 4. DATA UNDERSTANDING

#### 1. Dataset: The dataset used in this study is from Kaggle, https://www.kaggle.com/datasets/mnassrib/telecom-churn-datasets. The dataset contains 3333 observations with 20 variables. The variables include:

#### **State:** Categorical variable indicating the customer's state.
#### **Account length:** Numeric variable indicating the length of the customer account.
#### **Area code:** Numeric variable indicating the area code of the
#### **Phone number:** Categorical variable (likely to be excluded as it won't contribute to churn prediction).
#### **International plan:**  Categorical variable indicating if the customer has an international plan.
#### **Voice mail plan:** Categorical variable indicating if the customer has a voicemail plan.
#### **Number vmail messages:** Numeric variable indicating the number of voicemail messages.
#### **Total day minutes:** Numeric variables indicating usage minutes in various time segments.
#### **Total day calls:**
#### **Total day charge:**
#### **Total eve minutes:** Numeric variables indicating usage minutes in various time segments.
#### **Total eve calls:** 
#### **Total eve charge:**
#### **Total night minutes:** Numeric variables indicating usage minutes in various time segments.
#### **Total night calls:**
#### **Total night charge:**
#### **Total intl minutes:** Numeric variables indicating usage minutes in various time segments.
#### **Total intl calls:**
#### **Total intl charge:**
#### **Customer service calls:**
#### **Churn:** Binary target variable indicating customer churn (True/False).

#### The variables in the dataset will enable us to explore connection between various cosmuner habits and churn.


## 5. DATA PREPARATION

#### **Importing  Packages**

In [2]:
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.linear_model import LogisticRegression
import numpy as np

#### **EXPLORATORY DATA ANALYSIS**

**Loading Data**

In [3]:
import pandas as pd
df= pd.read_csv("Data/bigml_59c28831336c6604c800002a.csv", index_col=0)
print(df.head())
print("--------------------")
print(df.shape)

       account length  area code phone number international plan  \
state                                                              
KS                128        415     382-4657                 no   
OH                107        415     371-7191                 no   
NJ                137        415     358-1921                 no   
OH                 84        408     375-9999                yes   
OK                 75        415     330-6626                yes   

      voice mail plan  number vmail messages  total day minutes  \
state                                                             
KS                yes                     25              265.1   
OH                yes                     26              161.6   
NJ                 no                      0              243.4   
OH                 no                      0              299.4   
OK                 no                      0              166.7   

       total day calls  total day charge  total eve m

In [4]:
#drop the hone number column
df.drop("phone number",axis=1,inplace=True)

**Checking for Missing values**

The data set contains no missing values 

In [5]:
df.isna().sum()

account length            0
area code                 0
international plan        0
voice mail plan           0
number vmail messages     0
total day minutes         0
total day calls           0
total day charge          0
total eve minutes         0
total eve calls           0
total eve charge          0
total night minutes       0
total night calls         0
total night charge        0
total intl minutes        0
total intl calls          0
total intl charge         0
customer service calls    0
churn                     0
dtype: int64

 **Checking Data Types**

In [6]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 3333 entries, KS to TN
Data columns (total 19 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   account length          3333 non-null   int64  
 1   area code               3333 non-null   int64  
 2   international plan      3333 non-null   object 
 3   voice mail plan         3333 non-null   object 
 4   number vmail messages   3333 non-null   int64  
 5   total day minutes       3333 non-null   float64
 6   total day calls         3333 non-null   int64  
 7   total day charge        3333 non-null   float64
 8   total eve minutes       3333 non-null   float64
 9   total eve calls         3333 non-null   int64  
 10  total eve charge        3333 non-null   float64
 11  total night minutes     3333 non-null   float64
 12  total night calls       3333 non-null   int64  
 13  total night charge      3333 non-null   float64
 14  total intl minutes      3333 non-null   float6

**Defining X and Y variables**

In [7]:
X=df.drop("churn",axis=1)
y=df["churn"]

**Performing test train split**

In [8]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.5,random_state=42)

## 6. PREPROCESSING

 **Selecting categorical variables**

In [9]:
X_train_cat = X_train.select_dtypes(include="object")
X_train_num = X_train.select_dtypes(exclude="object")
X_test_cat = X_test.select_dtypes(include="object")
X_test_num = X_test.select_dtypes(exclude="object")

**One hot encoding**

In [10]:
Ohe = OneHotEncoder(sparse_output=False, handle_unknown="ignore")
X_train_cat_ohe =Ohe.fit_transform(X_train_cat)
X_test_cat_ohe =Ohe.transform(X_test_cat)

**Creating Dataframes**

In [11]:
X_train_cat_ohe_df= pd.DataFrame(X_train_cat_ohe, index=X_train_cat.index,columns= np.hstack(Ohe.categories_))
X_test_cat_ohe_df= pd.DataFrame(X_test_cat_ohe, index=X_test_cat.index,columns= np.hstack(Ohe.categories_))
X_train_cat_ohe_df.head()
X_test_cat_ohe_df.head()

no  yes   no  yes
state                    
WY     1.0  0.0  1.0  0.0
IL     1.0  0.0  1.0  0.0
SD     1.0  0.0  1.0  0.0
KY     1.0  0.0  1.0  0.0
WY     1.0  0.0  1.0  0.0

**Scaling the Numeric Variables**

In [12]:
from sklearn.preprocessing import MinMaxScaler
Scaler= MinMaxScaler()
X_train_num_scaled= Scaler.fit_transform(X_train_num)
X_test_num_scaled = Scaler.transform(X_test_num)


**Creating Final X_train Dataset**
- This step involves converting the both scaled and encoded data iinto dataframes to enable merging and creation of the final x_train dataset for modelling purposes

In [13]:
X_train_cat_ohe_df= pd.DataFrame(X_train_cat_ohe, index=X_train_cat.index,columns= np.hstack(Ohe.categories_))
X_train_num_scaled_df = pd.DataFrame(X_train_num_scaled, index=X_train_num.index, columns=X_train_num.columns)
X_test_cat_ohe_df= pd.DataFrame(X_test_cat_ohe, index=X_test_cat.index,columns= np.hstack(Ohe.categories_))
X_test_num_scaled_df = pd.DataFrame(X_test_num_scaled, index=X_test_num.index, columns=X_test_num.columns)
X_train_final_df = pd.concat([X_train_num, X_train_cat_ohe_df], axis=1)
X_test_final_df = pd.concat([X_test_num, X_test_cat_ohe_df], axis=1)
print(X_train_final_df.shape)
print(X_test_final_df.shape)

#print (X_train_cat_ohe_df.shape)
#print(X_train_num_scaled_df.shape)
#X_train_final_df = pd.concat([X_train_num, X_train_cat_ohe_df], axis=1)
#print(X_train_final_df.shape)
#drop the second row
#X_train_final_df.drop(X_train_final_df.index[1], inplace=True)
#_train_final_df.head()


(1666, 20)
(1667, 20)


#### 7. MODELLING

**Fitting the logistic regression model**

This step involves creating a model that uses the logistic regression algorithm to predict Churn among the telcom cleints. first we define/instantiate the model and then fit the training data.


In [14]:
#Instantiate the model
LogisticRegressionModel = LogisticRegression(fit_intercept=False, solver= 'liblinear',max_iter=2000)
#Fit the model
Baseline_model = LogisticRegressionModel.fit(X_train_final_df, y_train)

**Prediction**

This step involves making predictions on the test data using the model that we just trained.

In [15]:
y_train_pred = Baseline_model.predict(X_train_final_df)
y_test_pred = Baseline_model.predict(X_test_final_df)

#### 8. EVALUATION

In [17]:
from sklearn.metrics import precision_score, recall_score, accuracy_score, f1_score

print('Train precision: %.3f' % precision_score(y_train, y_train_pred))
print('Test precision: %.3f' % precision_score(y_test, y_test_pred))
print("-------------------------------------------")
print('Train recall: %.3f' % recall_score(y_train, y_train_pred))
print('Test recall: %.3f' % recall_score(y_test, y_test_pred))
print("-------------------------------------------")
print('Train accuracy: %.3f' % accuracy_score(y_train, y_train_pred))
print('Test accuracy: %.3f' % accuracy_score(y_test, y_test_pred))
print("-------------------------------------------")
print('Train f1: %.3f' % f1_score(y_train, y_train_pred))
print('Test f1: %.3f' % f1_score(y_test, y_test_pred))


Train precision: 0.568
Test precision: 0.579
-------------------------------------------
Train recall: 0.180
Test recall: 0.176
-------------------------------------------
Train accuracy: 0.866
Test accuracy: 0.857
-------------------------------------------
Train f1: 0.274
Test f1: 0.270


**Below is a brief description of the evaluation metrics of the baseline model:**

**Precision: Train: 0.568, Test: 0.579**

The precision values for both the train and test sets are similar, which suggests that the model has a consistent rate of correctly identifying positive instances across both sets. However, the precision is relatively low, indicating that a significant number of predicted positives are actually false positives.

**Recall: Train: 0.180 Test: 0.176**

The recall values are quite low for both train and test sets, which means the model is missing a large number of actual positive instances. This might suggest that the model is not very sensitive and is underfitting the positive class.

**Accuracy: Train: 0.866 Test: 0.857**

The accuracy is relatively high, but this might be misleading if the dataset is imbalanced (i.e., there are many more negatives than positives). High accuracy in such cases could simply reflect the model's ability to predict the majority class correctly.

**F1 Score: Train: 0.274 Test: 0.270**

The F1 score, which is the harmonic mean of precision and recall, is quite low for both train and test sets. This indicates a poor balance between precision and recall, reflecting overall weak performance in identifying the positive class.

#### 9. RECOMENDATIONS